modules/spring-config/src/main/java/org/mule/config/spring/SpringRegistry.java

```diff
@@ -38,76 +50,102 @@
 
+    private boolean readOnly;
+
+    private RegistrationDelegate registrationDelegate;
+
     //This is used to track the Spring context lifecycle since there is no way to confirm the
     //lifecycle phase from the application context
     protected AtomicBoolean springContextInitialised = new AtomicBoolean(false);
 
-    public SpringRegistry(MuleContext muleContext)
-    {
-        super(REGISTRY_ID, muleContext);
-    }
-
-    public SpringRegistry(String id, MuleContext muleContext)
-    {
-        super(id, muleContext);
-    }
-
     public SpringRegistry(ApplicationContext applicationContext, MuleContext muleContext)
     {
         super(REGISTRY_ID, muleContext);
-        this.applicationContext = applicationContext;
+        setApplicationContext(applicationContext);
     }
 
     public SpringRegistry(String id, ApplicationContext applicationContext, MuleContext muleContext)
     {
         super(id, muleContext);
-        this.applicationContext = applicationContext;
+        setApplicationContext(applicationContext);
     }
 
     public SpringRegistry(ConfigurableApplicationContext applicationContext, ApplicationContext parentContext, MuleContext muleContext)
     {
         super(REGISTRY_ID, muleContext);
         applicationContext.setParent(parentContext);
-        this.applicationContext = applicationContext;
+        setApplicationContext(applicationContext);
     }
 
     public SpringRegistry(String id, ConfigurableApplicationContext applicationContext, ApplicationContext parentContext, MuleContext muleContext)
     {
         super(id, muleContext);
         applicationContext.setParent(parentContext);
+        setApplicationContext(applicationContext);
+    }
+
+    private void setApplicationContext(ApplicationContext applicationContext)
+    {
         this.applicationContext = applicationContext;
+        if (applicationContext instanceof ConfigurableApplicationContext)
+        {
+            readOnly = false;
+            registrationDelegate = new ConfigurableRegistrationDelegate((ConfigurableApplicationContext) applicationContext);
+        }
+        else
+        {
+            readOnly = true;
+            registrationDelegate = new ReadOnlyRegistrationDelegate();
+        }
     }
 
     @Override
     protected void doInitialise() throws InitialisationException
     {
-        if (applicationContext instanceof ConfigurableApplicationContext)
+        if (!readOnly)
         {
-                ((ConfigurableApplicationContext) applicationContext).refresh();
+            ((ConfigurableApplicationContext) applicationContext).refresh();
         }
+
+        initTransformers();
         //This is used to track the Spring context lifecycle since there is no way to confirm the lifecycle phase from the application context
         springContextInitialised.set(true);
     }
 
+    private void initTransformers()
+    {
+        MuleRegistryHelper registryHelper = (MuleRegistryHelper) muleContext.getRegistry();
```

Comment:
This downcast and depending on an implementation instead of a interface is ugly. This also required to make registerTransformerResolver public in MuleRegistryHelper